service/controller/release/resource/argoapps/resource.go

```diff
@@ -0,0 +1,255 @@
+	}
+
+	return r, nil
+}
+
+func (r *Resource) Name() string {
+	return Name
+}
+
+func (r *Resource) ensureState(ctx context.Context) error {
+	var releases releasev1alpha1.ReleaseList
+	{
+		err := r.k8sClient.CtrlClient().List(
+			ctx,
+			&releases,
+		)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		releases = key.ExcludeDeletedRelease(releases)
+		releases = key.ExcludeUnusedDeprecatedReleases(releases)
+	}
+
+	var components map[string]releasev1alpha1.ReleaseSpecComponent
+	{
+		components = key.ExtractComponents(releases)
+	}
+
+	componentApps := map[string]argoappv1alpha1.Application{}
+	{
+		for name, component := range components {
+			argoApp, err := r.componentToArgoApplication(ctx, component)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			componentApps[name] = *argoApp
+		}
+	}
+
+	var apps argoappv1alpha1.ApplicationList
+	{
+		err := r.k8sClient.CtrlClient().List(
+			ctx,
+			&apps,
+			&client.ListOptions{
+				Namespace: argoCDNamespace,
+				LabelSelector: labels.SelectorFromSet(labels.Set{
+					key.LabelManagedBy: project.Name(),
+				}),
+			},
+		)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	appsToDelete := calculateObsoleteApps(componentApps, apps)
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("calculated %d obsolete apps", len(appsToDelete.Items)))
+	for i, app := range appsToDelete.Items {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting app %#q in namespace %#q", app.GetName(), app.GetNamespace()))
+
+		err := r.k8sClient.CtrlClient().Delete(
+			ctx,
+			&appsToDelete.Items[i],
+		)
+		if apierrors.IsNotFound(err) {
+			// fall through.
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted app %#q in namespace %#q", app.GetName(), app.GetNamespace()))
+	}
+
+	appsToCreate := calculateMissingApps(componentApps, apps)
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("calculated %d missing apps", len(appsToCreate.Items)))
+	for i, app := range appsToCreate.Items {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating app %#q in namespace %#q", app.GetName(), app.GetNamespace()))
+
+		// ensure giantswarm.io/managed-by label
+		if appsToCreate.Items[i].Labels == nil {
+			appsToCreate.Items[i].Labels = map[string]string{}
+		}
+		appsToCreate.Items[i].Labels[key.LabelManagedBy] = project.Name()
+
+		err := r.k8sClient.CtrlClient().Create(
+			ctx,
+			&appsToCreate.Items[i],
+		)
+		if apierrors.IsAlreadyExists(err) {
+			// fall through.
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("created app %#q in namespace %#q", app.GetName(), app.GetNamespace()))
+	}
+
+	return nil
+}
+
+func calculateMissingApps(componentApps map[string]argoappv1alpha1.Application, apps argoappv1alpha1.ApplicationList) argoappv1alpha1.ApplicationList {
```

Comment:
Maybe I am blind but there seem to be no unit tests for any of these functions. We previously had coverage for the equivalent functions and I would like to keep that coverage.

This is still a quite critical part of our deployments right now.

Reply:
No, you're not blind :joy: I'll add unit tests.